File: src/algorithms/strings/transformation/reverse-words/sources/reverse-words.ts
// Reverse Words in a String — split, reverse word order, rejoin with single spaces.
// Trims leading/trailing whitespace and collapses multiple spaces between words.
// Time: O(n)  Space: O(n)

function reverseWords(text: string): string {
  const words = text.trim().split(/\s+/); // @step:initialize

  let leftIndex = 0; // @step:initialize
  let rightIndex = words.length - 1; // @step:initialize

  while (leftIndex < rightIndex) {
    const leftWord = words[leftIndex]; // @step:read-char
    const rightWord = words[rightIndex]; // @step:read-char

    words[leftIndex] = rightWord ?? ""; // @step:swap-pointers
    words[rightIndex] = leftWord ?? ""; // @step:swap-pointers

    leftIndex++; // @step:visit
    rightIndex--; // @step:visit
  }

  return words.join(" "); // @step:complete
}

File: src/algorithms/strings/transformation/reverse-words/sources/reverse-words.go
// Reverse Words in a String — split, reverse word order, rejoin with single spaces.
// Trims leading/trailing whitespace and collapses multiple spaces between words.
// Time: O(n)  Space: O(n)

package main

import "strings"

func reverseWords(text string) string {
	words := strings.Fields(text) // @step:initialize

	leftIndex := 0 // @step:initialize
	rightIndex := len(words) - 1 // @step:initialize

	for leftIndex < rightIndex {
		leftWord := words[leftIndex] // @step:read-char
		rightWord := words[rightIndex] // @step:read-char

		words[leftIndex] = rightWord // @step:swap-pointers
		words[rightIndex] = leftWord // @step:swap-pointers

		leftIndex++ // @step:visit
		rightIndex-- // @step:visit
	}

	return strings.Join(words, " ") // @step:complete
}
